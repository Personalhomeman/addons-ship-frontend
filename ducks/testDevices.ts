import { Dispatch } from 'redux';
import { createAction, createReducer } from 'deox';

import api from '@/services/bitrise-api';
import { actionTypeCreator } from '@/utils';
import { TestDevice } from '@/models/test-device';

const $ = actionTypeCreator('TEST_DEVICES');

const _fetchTestDevices = (appSlug: string) => async (dispatch: Dispatch) => {
  dispatch(fetchTestDevices.next());

  try {
    const testDevices = await api.getTestDevices(appSlug);

    dispatch(fetchTestDevices.complete(testDevices));
  } catch (error) {
    dispatch(fetchTestDevices.error(error));
  }
};

export const fetchTestDevices = Object.assign(_fetchTestDevices, {
  next: createAction($`GET_NEXT`),
  complete: createAction($`GET_COMPLETE`, resolve => (testDevices: TestDevice[]) => resolve(testDevices)),
  error: createAction($`GET_ERROR`, resolve => (error: Error) => resolve(error))
});

const defaultState: TestDevicesState = [];

export type TestDevicesState = TestDevice[];
export const testDevices = createReducer(defaultState, handleAction => [
  handleAction(fetchTestDevices.complete, (_, { payload }) => payload)
]);
