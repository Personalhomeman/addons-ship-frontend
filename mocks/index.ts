import { AppVersion } from '@/models';
import { TestDevice } from '@/models/test-device';

export const mockAppVersion: AppVersion = {
  id: 123,
  version: '1.0.3',
  platform: 'ios',
  buildNumber: 32,
  buildSlug: '13245ads',
  lastUpdate: new Date('2018-09-22T12:42:31Z'),
  description: 'Some semi-long description',
  whatsNew: 'Some longer whats-new',
  minimumOs: '10.2',
  minimumSdk: 'dunno',
  packageName: 'com.package.name',
  certificateExpires: new Date('2018-09-22T12:42:31Z'),
  bundleId: 'bundle-id',
  size: 4567034,
  supportedDeviceTypes: ['iphone', 'ipad'],
  distributionType: 'development',
  iconUrl: 'http://placekitten.com/160/160',
  appName: 'Standup Timer',
  promotionalText: 'Promotional Text',
  keywords: 'Keywords',
  reviewNotes: 'Review Notes',
  supportUrl: 'https://bitrise.io/support',
  marketingUrl: 'https://bitrise.io/marketing',
  scheme: 'piramid',
  configuration: 'canary',
  publicInstallPageURL: 'https://bitrise.io/app/8b334705d8e78276'
};

export const mockTestDevices: TestDevice[] = [
  { deviceId: 'some-device-id-01', deviceType: 'ios', owner: 'test-user-1' },
  { deviceId: 'some-device-id-02', deviceType: 'android', owner: 'test-user-1' },
  { deviceId: 'some-device-id-03', deviceType: 'ios', owner: 'test-user-2' }
];
