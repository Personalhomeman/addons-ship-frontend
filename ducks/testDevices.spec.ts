jest.mock('@/services/ship-api');

import configureMockStore, { MockStoreCreator, MockStoreEnhanced } from 'redux-mock-store';
import thunk from 'redux-thunk';

import shipApi from '@/services/ship-api';

import { testDevices, fetchTestDevices } from './testDevices';
import { mockTestDevices } from '@/mocks';

describe('appVersion', () => {
  let mockStore: MockStoreCreator;
  beforeEach(() => {
    mockStore = configureMockStore([thunk.withExtraArgument({ shipApi })]);
  });

  describe('reducer', () => {
    it('loads devices', () => {
      const state = testDevices(undefined, fetchTestDevices.complete({ list: mockTestDevices, isLoading: false }));

      expect(state).toMatchSnapshot();
    });
  });

  describe('fetchTestDevices', () => {
    let store: MockStoreEnhanced;

    beforeEach(() => {
      store = mockStore();
    });

    it('fetches test devices', async () => {
      await store.dispatch(fetchTestDevices('app-slug') as any);

      expect(store.getActions()).toMatchSnapshot();
    });

    it("can't fetch test devices", async () => {
      (shipApi.getTestDevices as jest.Mock).mockRejectedValueOnce('api had some issue');
      await store.dispatch(fetchTestDevices('app-slug') as any);

      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
