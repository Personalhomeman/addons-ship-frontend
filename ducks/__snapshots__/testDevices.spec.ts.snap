// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appVersion fetchTestDevices can't fetch test devices 1`] = `
Array [
  Object {
    "type": "TEST_DEVICES_GET_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "TEST_DEVICES_GET_ERROR",
  },
]
`;

exports[`appVersion fetchTestDevices fetches test devices 1`] = `
Array [
  Object {
    "type": "TEST_DEVICES_GET_NEXT",
  },
  Object {
    "type": "TEST_DEVICES_GET_COMPLETE",
  },
]
`;

exports[`appVersion reducer loads devices 1`] = `
Object {
  "isLoading": false,
  "list": Array [
    Object {
      "deviceId": "some-device-id-01",
      "deviceType": "ios",
      "owner": "test-user-1",
    },
    Object {
      "deviceId": "some-device-id-02",
      "deviceType": "android",
      "owner": "test-user-1",
    },
    Object {
      "deviceId": "some-device-id-03",
      "deviceType": "ios",
      "owner": "test-user-2",
    },
  ],
}
`;
